### PR TITLE
core: Added simplified SEP-31 calculation

### LINKED ISSUE #236: Simply SEP-31 amount and fee calculation.

### PR DESCRIPTION
**Pull request by lijamie98**
Closes #236 

<details>
  <summary>PR Checklist</summary>

### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

This PR simplifies SEP-31 calculation according to i[ssue#236.](https://github.com/stellar/java-stellar-anchor-sdk/issues/236) 

### Why

See [issue#236](https://github.com/stellar/java-stellar-anchor-sdk/issues/236)

### Known limitations
N/A

### CODE REVIEW

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 1]
```diff
@@ -148,6 +148,55 @@ void updateAmounts() throws AnchorException {
...
+    // Update transactions
+    txn.setAmountFee(String.valueOf(fee));
+    txn.setAmountFeeAsset(feeAsset);
+    txn.setAmountIn(String.valueOf(amountIn));
+    txn.setAmountInAsset(amountInAsset);
+    txn.setAmountOut(String.valueOf(amountOut));
+    txn.setAmountOutAsset(amountOutAsset);
+  }
+
+  void updateAmounts2() throws AnchorException {
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:196]: I think there could be 3 methods here:
* updateAmountsWithoutQuote
* updateAmountsWithQuoteAndEmbeddedFees
* updateAmountsWithQuoteAndNonEmbeddedFees

WDYT?

Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:199]: 2 scenarios should be included in the following block. For `EmbeddedFees` and `NonEmbeddedFeeds`, we may discuss in Slack.

```java
    if (withoutQuote) {
      boolean strictSend = sep31Config.getPaymentType() == STRICT_SEND;
      // Without quote.
      amountOutAsset = amountInAsset;
      if (strictSend) {
        // amount_in = req.amount
        // amount_out = amount_in - amount fee
        amountIn = reqAmount;
        amountOut = amountIn.subtract(fee);
      } else {
        // amount_in = req.amount + fee
        // amount_out = req.amount
        amountIn = reqAmount.add(fee);
        amountOut = reqAmount;
      }
    } else {
      // With quote
      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
      txn.setQuoteId(request.getQuoteId());
      amountOutAsset = quote.getBuyAsset();
      amountIn = decimal(quote.getSellAmount(), reqAsset);
      amountOut = decimal(quote.getBuyAmount(), reqAsset);
    }
```

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 2]
```diff
@@ -151,6 +151,55 @@ void updateAmounts() throws AnchorException {
...
+        amountOut = amountIn.subtract(fee);
+      } else {
+        // amount_in = req.amount + fee
+        // amount_out = req.amount
+        amountIn = reqAmount.add(fee);
+        amountOut = reqAmount;
+      }
+    } else {
+      // With quote
+      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:183]: You need to handle the case where the quote can't be found. What if a person submits a transaction with a non-existing quote, right?

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 3]
```diff
@@ -151,6 +151,55 @@ void updateAmounts() throws AnchorException {
...
+      txn.setQuoteId(request.getQuoteId());
+      amountOutAsset = quote.getBuyAsset();
+      amountIn = decimal(quote.getSellAmount(), reqAsset);
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:186]: I think you should also validate that the PostTransaction amount is equals to the quote's sellAmount. If not, return a 400.

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 4]
```diff
@@ -151,6 +151,55 @@ void updateAmounts() throws AnchorException {
...
+      amountOut = decimal(quote.getBuyAmount(), reqAsset);
+    }
+
+    // Update transactions
+    txn.setAmountFee(String.valueOf(fee));
+    txn.setAmountFeeAsset(feeAsset);
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:192]: What to do if the quote price details are complex and contain multiple assets? You mentioned to let it empty?

**Review by marcelosalloum — APPROVED** [review]
Feel free to merge it after addressing the feedback.

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java (hunk 1 above)
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:199]: Oh, I got it. 

Do we plan to support both calculation? My understanding is that we are going to keep one of them and remove the other one.

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java (hunk 2 above)
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:183]: The validations are done before it enters the `updateAmount()` function. 😺 

The case is validated here at [Line#374](https://github.com/stellar/java-stellar-anchor-sdk/blob/2022-05-09-sep31-simplify-amount-fee/core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java#L374).

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java (hunk 3 above)
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:186]: I am not sure about this when I implement this. I actually posted the comment at [here](https://github.com/stellar/stellar-protocol/pull/1199#discussion_r870403310). 

```java
    request.amount = 100
    sep31Transaction.amount_in = 102
    sep31Transaction.amount_in_asset = USDC
    sep31Transaction.amount_out = 500
    sep31Transaction.amount_out_asset = BRL
    quote.sell_amount=102
    quote.sell_amount_asset=USDC
    quote.buy_amount=500
    quote.buy_amount_asset=BRL
```

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java (hunk 4 above)
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:192]: Cool! Good catch! 

Fixed.

**Comment by lijamie98** [comment]
Most of the feedback had been taken care of except extracting the `withoutQuote` function. I will merge this PR first. If we decide to support both calculation, I will work on the extraction.

**Pull request merged**